lib/getConfigForFile.js

```diff
@@ -57,19 +58,52 @@ module.exports = async function getConfigForFile(
 		stopDir: STOP_DIR,
 	});
 
-	let config = stylelint._options.configFile
-		? await configExplorer.load(stylelint._options.configFile)
-		: await configExplorer.search(searchPath);
+	let config = configFile
+		? await configExplorer.load(configFile)
+		: await configExplorer.search(opts.searchPath);
 
 	if (!config) {
 		config = await configExplorer.search(cwd);
 	}
 
 	if (!config) {
 		return Promise.reject(
-			configurationError(`No configuration provided${searchPath ? ` for ${searchPath}` : ''}`),
+			configurationError(
+				`No configuration provided${opts.searchPath ? ` for ${opts.searchPath}` : ''}`,
+			),
 		);
 	}
 
 	return config;
 };
+
+/**
+ * @param {StylelintInternalApi} stylelint
+ * @param {string} [searchPath]
+ * @param {string} [filePath]
+ * @returns {Promise<StylelintCosmiconfigResult>}
+ */
+module.exports = async function getConfigForFile(
+	stylelint,
+	searchPath = stylelint._options.cwd,
+	filePath,
+) {
+	const { config, mergeRules } = stylelint._options;
+
+	if (config && !mergeRules) {
+		return getAugmentedConfig(stylelint, filePath);
+	}
+
+	const cosmiconfigResult = await createCosmiconfig(stylelint, { filePath, searchPath });
+
+	if (config?.rules && mergeRules) {
```

Comment:
This file has been refactored to improve readability.
The only relevant change is this `if` and the one at l93.